Add translations documentation (#35)
front-commerce/developers.front-commerce.com: source/docs/advanced/theme/translations.md

---
id: translations
title: Translate your application
---

> Big e-commerce websites often need to translate their content in several languages to make sure that they reach a broader range of customers.

Reviewer: I am not a fan of this sentence, but it’s difficult to point out what makes me feel like that exactly. Here are suggestions to express ideas, we could discuss it to find a correct one if you want.

```suggestion
International e-commerce websites often need to translate their UI in several languages to make sure that they reach a broader range of customers.
```

or

```suggestion
Global e-commerce websites often need to allow users to choose their preferred language to make sure that they reach a broader range of customers.
```

In this documentation, we will see how this works for static content in your application.

<blockquote class="note">

> If you want to configure your application to support multiple languages, please refer to [Configure multiple stores](/docs/advanced/production-ready/multistore.html).

Reviewer: ```suggestion
Often you may want to change dynamic content (CMS pages, product descriptions…) in addition to translating the UI. If you want to configure your application to support multiple languages, please refer to [Configure multiple stores](/docs/advanced/production-ready/multistore.html).
```

Reviewer: I saw it was also mentioned in the last section, so feel free to discard this suggestion if you’d like.

</blockquote>

> Front-Commerce manages your translations by using [`react-intl`](https://github.com/yahoo/react-intl), a standard in the React ecosystem. This library works by transforming your static values in React Components. These components will then fetch your translations and display them in your application.

Reviewer: ```suggestion
Front-Commerce manages your translations by using [`react-intl`](https://github.com/yahoo/react-intl), a standard library in the React ecosystem. This library works by transforming your static values in React Components. These components will then fetch your translations and display them in your application.
```

## Declare translations in your application

For instance, let's see how to transform your values in react-intl Components.

* Strings
```diff
- Default translation
+ import { FormattedMessage } from "react-intl";
+ <FormattedMessage
+   id="My.Component.Id.messageId"
+   defaultMessage="Default translation"
+ />
```
* Dates
```diff
- import moment from "moment";
- moment(date).format("DD/MM/YYYY");
+ <FormattedDate
+   value={props.date}
+   year="numeric"
+   month="long"
+   day="2-digit"
+ />
```
* Prices
```diff
- <span>{value.amount} €</span>
+ import { FormattedNumber } from "react-intl";
+ <FormattedNumber
+   value={value.amount}
+   style="currency"
+   currency={value.currency}
+ />
```
    <blockquote class="note">

>     However in this particular we have abstracted this in Front-Commerce with the [`theme/components/atoms/Price`](https://gitlab.com/front-commerce/front-commerce/tree/develop/src/web/theme/components/atoms/Typography/Price) component.

Reviewer: ```suggestion
    However, for this particular use case we have abstracted this in Front-Commerce with the [`theme/components/atoms/Price`](https://gitlab.com/front-commerce/front-commerce/tree/develop/src/web/theme/components/atoms/Typography/Price) component (and its variants: `ProductPrice`, `PriceOrFree`). We recommend you to use it instead, so you could benefit from better integration with its related GraphQL Type.
```

    </blockquote>
* [and many more](https://github.com/yahoo/react-intl/wiki/Components)

### Without React Component

These components will be enough in most cases. However, the issue with these is that it wraps your string within a React Element, which may be troublesome when you actually need to handle the string itself.

For instance, this is the case when you want to add some label attributes to your DOM elements.

```jsx
<span class="icon" aria-label="Icon title displayed for screen readers" />
```

> Fortunatly, this is correctly handled by react-intl if you use [`defineMessages`](https://github.com/yahoo/react-intl/wiki/API#definemessages) combined with [`injectIntl`](https://github.com/yahoo/react-intl/wiki/API#injection-api).

Reviewer: ```suggestion
Fortunately, this is correctly handled by react-intl if you use [`defineMessages`](https://github.com/yahoo/react-intl/wiki/API#definemessages) combined with [`injectIntl`](https://github.com/yahoo/react-intl/wiki/API#injection-api).
```

<blockquote class="wip">
Please refer to [react-intl documentation](https://github.com/yahoo/react-intl/wiki/API#injection-api) for more information about these two methods. We plan to add an example here in the future but didn't have time to document it yet.
</blockquote>

> ## Setup your translations what's in your components

Reviewer: I really do not see what you wanted to write here… but there definitely seems to be a mistake!

> Now that you have defined what should be translated in your application, you actually need to translate your application. This is doable in two steps:

Reviewer: ```suggestion
Now that you have defined what should be translated in your application, you actually need to translate your application. This is a two-step process:
```

```sh
npm run translate
```
2. Translate the strings available in those files, and you are good to go :)

Note that for some translations you won't need to change anything. This usually happens for the default language and for business phrases such as "SKU", etc.

However, react-intl will warn you against those cases because it might be an actual translation that is missing. Luckily, you can opt-out of this warning for specific translations ids by adding it to the `my-module/translations/locales/whitelist_[lang].json` file.

This task can be daunting at first, but it will make your life easier in the future when you will need to know which string was not translated in a recent update.

## About dynamic content

`react-intl` lets you translate your static content. But what about dynamic content? Things like product title and description, CMS pages, etc.

This is done on the GraphQL side. The principle is that when a user is connected to Front-Commerce, they will be assigned a `storeViewCode` in their session (see [Configure multiple stores](/docs/advanced/production-ready/multistore.html) for more details).

> This code will then be used by your GraphQL loaders to retreive the correct content from your backend. Learn more about [GraphQL loaders](/docs/advanced/graphql/slim-down-resolvers-with-loaders.html).

Reviewer: ```suggestion
This code will then be used by your GraphQL loaders to retrieve the correct content from your backend. Learn more about [GraphQL loaders](/docs/advanced/graphql/slim-down-resolvers-with-loaders.html).
```
